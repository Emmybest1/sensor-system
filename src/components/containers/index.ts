export * from './error-boundary/error-boundary.component';
