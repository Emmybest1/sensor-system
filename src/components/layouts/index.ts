export * from './main/main.component';
export * from './footer/footer.component';
