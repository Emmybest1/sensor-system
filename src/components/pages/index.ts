export { default as Login } from './login/login.component';
export { default as SignUp } from './signup/signup.component';
export { default as Fallback } from './fallback/fallback.component';
