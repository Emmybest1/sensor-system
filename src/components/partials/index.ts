export * from './logo/logo.component';
export * from './input/input.component';
export * from './button/button.component';
