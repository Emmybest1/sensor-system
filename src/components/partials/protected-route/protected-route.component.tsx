import React from 'react';
import { Route, RouteComponentProps, useHistory } from 'react-router-dom';

import { Paths } from 'routes';
import { getCurrentUser } from 'services';
import { checkIsUserAuthorizedToAccessPage } from 'utils';

type ProtectedRoutePropTypes = {
  pagePermissionNumber: number;
  [key: string]: any;
};

const ProtectedRoute: React.FC<ProtectedRoutePropTypes> = ({ pagePermissionNumber, component: Component, ...restProps }) => {
  const history = useHistory();

  const currentUser = getCurrentUser();

  const isAuthorized = checkIsUserAuthorizedToAccessPage(pagePermissionNumber, currentUser?.permissions);

  if (!isAuthorized) {
    history.push({
      pathname: Paths.fallback,
      state: {
        heading: 'Not Authorized',
        message: 'You are not authorized to access this page.',
        shownButton: true,
        fallbackPath: Paths.login,
      },
    });
    return null;
  }

  return <Route {...restProps} render={(props: RouteComponentProps) => <Component {...restProps} {...props} />} />;
};

export { ProtectedRoute };
