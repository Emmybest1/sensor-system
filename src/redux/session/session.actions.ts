/* eslint-disable no-unused-vars */
import { action } from 'typesafe-actions';
import { ActionCreator, Action } from 'redux';

import { Paths } from 'routes';
import { PAGES_PERMISSIONS, User } from './session';
import { types } from './session.types';
import { $senTemApi, LoginData } from 'api';
import { CurrentUser, storeCurrentUser } from 'services';

/** *****
 * login actions creators
 */
const loginSuccessful: ActionCreator<Action<string>> = (payload: User) => action(types.LOG_IN_SUCCESS, payload);

const loginFailed: ActionCreator<Action<string>> = (payload: string) => action(types.LOG_IN_FAILURE, payload);

const resetLoginError: ActionCreator<Action<string>> = () => action(types.RESET_LOG_IN_FAILURE);

const login = (data: LoginData, onSuccessDispatch?: Function, onErrorDispatch?: Function) => (dispatch: (arg0: Action<string>) => void) => {
  dispatch(action(types.LOG_IN_STARTED));

  $senTemApi
    .login(`${process.env.REACT_APP_OPENDATA_DATABASE_URL}api/login`, data)
    .then((response) => {
      const currentUser: CurrentUser = Object.assign({ permissions: [PAGES_PERMISSIONS] }, response) as unknown as CurrentUser;

      storeCurrentUser(Object.create(currentUser));

      if (onSuccessDispatch) onSuccessDispatch();

      window.location.replace(Paths.home);

      dispatch(loginSuccessful(response as User));
    })
    .catch((error) => {
      if (onErrorDispatch) onErrorDispatch();

      dispatch(loginFailed(error as string));
    });
};

/** *****
 *  signup actions creators
 */
const signupSuccessful: ActionCreator<Action<string>> = (payload: User) => action(types.LOG_IN_SUCCESS, payload);

const signupFailed: ActionCreator<Action<string>> = (payload: string) => action(types.LOG_IN_FAILURE, payload);

const resetSignupError: ActionCreator<Action<string>> = () => action(types.RESET_LOG_IN_FAILURE);

const signup = (data: LoginData, onSuccessDispatch?: Function, onErrorDispatch?: Function) => (dispatch: (arg0: Action<string>) => void) => {
  dispatch(action(types.LOG_IN_STARTED));

  $senTemApi
    .login(`${process.env.REACT_APP_OPENDATA_DATABASE_URL}api/login`, data)
    .then((response) => {
      const currentUser: CurrentUser = Object.assign({ permissions: [PAGES_PERMISSIONS] }, response) as unknown as CurrentUser;

      storeCurrentUser(Object.create(currentUser));

      if (onSuccessDispatch) onSuccessDispatch();

      window.location.replace(Paths.home);

      dispatch(loginSuccessful(response as User));
    })
    .catch((error) => {
      if (onErrorDispatch) onErrorDispatch();

      dispatch(signupFailed(error as string));
    });
};

export { login, resetLoginError, signup, resetSignupError };
