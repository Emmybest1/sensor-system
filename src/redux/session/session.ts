import { ActionType } from 'typesafe-actions';

import { GenericState, GENERIC_STATE_KEYS } from 'redux/root';

export interface User {
  id: number | string;
  email: string;
  accessToken: string;
}

export interface SessionState extends GenericState {
  user: User;
}
export type SessionActions = ActionType<typeof import('./session.actions')>;

const SESSION_NAME = 'session';

const SESSION_STATE_KEYS = {
  ...GENERIC_STATE_KEYS,
  user: 'user',
};

const PAGES_PERMISSIONS = {
  HOME: 1,
  HISTORIES: 2,
  HISTORY: 3,
};

export { SESSION_NAME, SESSION_STATE_KEYS, PAGES_PERMISSIONS };
