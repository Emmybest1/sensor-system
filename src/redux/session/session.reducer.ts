import { Reducer } from 'redux';

import { types } from './session.types';
import { User, SessionState } from './session';

const initialState: SessionState = {
  isLoading: false,
  error: null,
  user: {} as User,
};

const sessionReducer: Reducer<SessionState, SessionState | any> = (state = initialState, action): SessionState => {
  switch (action.type) {
    case types.LOG_IN_STARTED: {
      return { ...state, isLoading: true };
    }

    case types.LOG_IN_SUCCESS: {
      return { ...state, isLoading: false, error: null, user: action.payload as User };
    }

    case types.LOG_IN_FAILURE: {
      return { ...state, isLoading: false, error: action.payload as string };
    }

    case types.RESET_LOG_IN_FAILURE: {
      return { ...state, error: null };
    }

    case types.SIGN_UP_STARTED: {
      return { ...state, isLoading: true };
    }

    case types.SIGN_UP_SUCCESS: {
      return { ...state, isLoading: false, error: null, user: action.payload as User };
    }

    case types.SIGN_UP_FAILURE: {
      return { ...state, isLoading: false, error: action.payload as string };
    }

    case types.RESET_SIGN_UP_FAILURE: {
      return { ...state, error: null };
    }

    case types.LOG_OUT_SUCCESS: {
      return initialState;
    }

    default: {
      return state;
    }
  }
};

export { sessionReducer };
