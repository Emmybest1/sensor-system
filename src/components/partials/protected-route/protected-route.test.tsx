import { render } from '@testing-library/react';
import { BrowserRouter as Router } from 'react-router-dom';

import { Paths } from 'routes';
import { Home } from 'components/pages';
import { getCurrentUser } from 'services';
import { PAGES_PERMISSIONS } from 'redux/session/session';
import { ProtectedRoute } from './protected-route.component';

describe('<ProtectedRouter/>', () => {
  test('should render Component', () => {
    const { container } = render(
      <Router>
        <ProtectedRoute exact path={Paths.event} component={Home} pagePermissionNumber={PAGES_PERMISSIONS.HOME} />
      </Router>
    );
    const currentUser = getCurrentUser();

    if (currentUser) {
      expect(window.location.pathname).toEqual(Paths.home);
    } else {
      expect(window.location.pathname).toEqual(Paths.fallback);
    }

    expect(container.firstChild).toMatchSnapshot();
  });
});
