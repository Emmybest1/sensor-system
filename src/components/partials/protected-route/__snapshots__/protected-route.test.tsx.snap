// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProtectedRouter/> should render Component 1`] = `null`;
